lib/spack/spack/test/versions.py

```diff
@@ -621,7 +621,7 @@ def test_git_hash_comparisons(
     # Spec based on earliest commit
     spec0 = spack.spec.Spec('git-test-commit@%s' % commits[-1])
     spec0.concretize()
-    assert spec0.satisfies('@:0')
+    assert spec0.satisfies('@999:')
```

Comment:
This commit should be detected as < 1.0... but I'm not sure how best to make this work.

I think the problem is that it's very difficult to distinguish between "I don't know the relationship between this commit and versions" and "I know this commit comes before any version". This commit should be the latter, and only the former should change to be unknown versions.

Maybe for versions with `ancestor_commits = []` we can check whether it's an ancestor of the earliest known commit?

Reply:
Ok, I probably misread the test.  We could make it forward-search to determine if it's before a commit to somewhat take care of that, but that's a larger and slightly harder change.  This is also a change in behavior from before, it was always treating it as "I have no idea what this is", but it happened to work because that produced a minimum version.  I'll take a look and see if I can reasonably easily hack something in to give us a "this is before the first" or something and give it a min version of some sort.

Reply:
Just pushed an update for this.  The logic now searches backward first, then if nothing is found it does the same search in reverse.  If at least one commit is found, it uses a string version that states how it was derived, so it's effectively a minimum version.  It really should create a version below the found version by a distance, but there's no good way to do that with the version scheme as it stands.  At least it represents "this is below some version" a bit better, but it can't yet provide any more information than being low.  What do you think @becker33?